### ISSUE TITLE
Drag and drop an image in a comment doesn't work (Enterprise)

### ISSUE THREAD

**Issue by patdolem** [huboard/huboard#520]
It allows me to drag and drop an image but it's not adding the url in the comment:
![huboard](https://cloud.githubusercontent.com/assets/5685914/6474975/81353958-c1ba-11e4-99bc-179a08775625.jpg)
Instead it opens a new page that includes the image:
![huboard2](https://cloud.githubusercontent.com/assets/5685914/6475042/21063d6a-c1bb-11e4-9687-96955bd3bf45.jpg)

**Comment by patdolem** [comment]
Related to #395 (closed issue)

**Comment by patdolem** [comment]
This is still not working and prevent us to add any image without having to go through GitHub Enterprise.

**Comment by chapati23** [comment]
still not working. tbh this is a _very_ big blocker for using huboard as a real backlog- and sprint-tool for us. PO can't upload images and has to very awkwardly upload them somewhere else and then put links into the ticket description

**Comment by dahlbyk** [comment]
Can you please confirm that image upload is actually enabled? The comment box would have a footer that also allows manual upload:
<img width="415" alt="firefox_2016-05-19_09-30-43" src="https://cloud.githubusercontent.com/assets/133987/15397101/798c3732-1da4-11e6-993f-53e846439b53.png">

In @patdolem's screenshot above, that footer is missing, so it doesn't look like upload is enabled. Assuming that's the case, we should fix that the UI implies the drag-and-drop will work even though we don't have anywhere to save the image.

I'm just trying to clarify if this is actually a bug in the HB:E upload process, or just confusing UI.

-----

@chapati23 if upload _isn't_ enabled, let's figure out what's blocking that so your PO is able to work use HuBoard more effectively!

**Comment by patdolem** [comment]
@dahlbyk I just checked and the comment box doesn't have the footer you mention.

**Comment by chapati23** [comment]
@dahlbyk thanks for the quick reply.

where would i enable image upload? don't see it in my settings. probably need to be an admin?

**Comment by rauhryan** [comment]
@patdolem @chapati23 

Currently we disable image upload in Enterprise because there isn't an API to store assets in GitHub Enterprise *yet*

We've kicked around a few ideas on how to best enable it, but there are trade-offs with the approaches

 1. We can possibly store the image in a gist, we haven't tried doing this yet, but it's possible that we could create a gist with just the image and then use that url as the asset url, it would be cool delete the gist afterwards and the image still be there

 2. Store and serve images directly on the HuBoard appliance. Tradeoff: your images are now dependent on HuBoard being up and available and complicated things if you need to move HuBoard to a different domain or in the worse case decommission the appliance. 

 3. Store the assets in an S3 bucket that you own. 

We already support option 3, but it takes some setup and of course there are tradeoffs with this approach as well. 

Ideally we would store images in GitHub:Enterprise, so that they are managed along side the images that are uploaded with your normal issues. 

As an integrator this has been our top request to the GitHub API team and for whatever reason they don't really have plans on adding it as far as we know. 

It would go a long way to send a support request to GitHub for upload support or talk to your sales rep. In the meantime, we support AWS S3 and I'm happy to help you get it set up if you need it.

**Comment by patdolem** [comment]
@rauhryan Thanks for this information. Just to clarify, is that meaning another trade-off will be to not have the image viewable on the issue in GitHub? If this is the case, we prefer to wait until GitHub Enterprise provide the api.

**Comment by dahlbyk** [comment]
Images being viewable from GitHub would be a requirement for any solution we would consider shipping. In the three scenarios listed above, embedded links would look something like:

1. `https://github.corp/gist/{user}/5350320500d7a22eb6f2fa36c4b0c7bf/my-image.jpg`, visible to anyone authenticated to `github.corp`
2. `https://huboard.corp/uploads/5350320500d7a22eb6f2fa36c4b0c7bf/my-image.jpg`, visible to anyone authenticated to `huboard.corp` (or to anonymous users with access to `huboard.corp`, if that's the preferred behavior)
3. `https://s3-zone.amazonaws.com/huboard.corp/uploads/5350320500d7a22eb6f2fa36c4b0c7bf/my-image.jpg`, visible to anyone with the URL

**Comment by dahlbyk** [comment]
Another option that has been brought up in internal discussion would be to "hide" HuBoard uploads inside the corresponding repository using some Git magic. This idea is actually several months old, and seem like it would work really well for users with Write access. However, users without Write access present a problem. Two options:

1. Fork the repository and store uploads there. But what happens if their account is disabled or their fork is deleted?
2. Create a `huboard` "owner" account that would be granted write access organization-wide as part of HB:E setup. Uploads from users without Write access would be saved into the repo by `huboard` on their behalf. The downside here is `huboard` taking up a GH:E seat. (Or, really, we could let you provide an access token for *any* user that would have global Write access - doesn't technically need to be HuBoard-specific.)

Thoughts?

**Comment by chapati23** [comment]
hmm, seems like a decent workaround to me. another advantage of having the assets inside your repo: it'd be easy to migrate once (or if…) GitHub enables uploading in their enterprise API. migration script of iterating over all assets and pushing them to GitHub Enterprise API wouldn't be too complicated, i guess.

**Comment by patdolem** [comment]
This is the answer I've got from our GitHub Admin:
Least bad option would be “we could let you provide an access token for any user that would have global Write access”, but that doesn’t sound to appealing either.

**Comment by rauhryan** [comment]
I looked into using Gists and it's a no go right now. 

The HTTP API doesn't support binary files, so I'm going to look into the content API and see if I can get an image into a repo as a file. 

TL;DR; Gists won't work without using the git protocol and for that we would need ssh keys

**Comment by dahlbyk** [comment]
> another advantage of having the assets inside your repo: it'd be easy to migrate once (or if…) GitHub enables uploading in their enterprise API. migration script of iterating over all assets and pushing them to GitHub Enterprise API wouldn't be too complicated, i guess.

Yeah, it would be somewhat tedious to iterate over all issues and comments in search of HuBoard-embedded links, but doable. Shouldn't really hurt anything to leave the content hidden in the Git repo, though, so maybe not worth the effort?

> This is the answer I've got from our GitHub Admin:
> Least bad option would be “we could let you provide an access token for any user that would have global Write access”, but that doesn’t sound to appealing either.

@patdolem for what it's worth, the code using such a token would be 100% open source for your review. We really don't want this level of access any more than you do, but our options are few.

I suppose another option would be something like https://owncloud.com/. Is anyone using some sort of on-prem file storage solution with an API?

**Comment by patdolem** [comment]
@dahlbyk Would such a solution allow to view the image in GitHub also?

**Comment by dahlbyk** [comment]
> @dahlbyk Would such a solution allow to view the image in GitHub also?

Yes. Any solution that doesn't allow viewing in both GitHub and HuBoard would seem to be a non-starter.

**Comment by dahlbyk** [comment]
> Another option that has been brought up in internal discussion would be to "hide" HuBoard uploads inside the corresponding repository using some Git magic. This idea is actually several months old, and seem like it would work really well for users with Write access.

FYI, this is in progress (https://github.com/huboard/huboard-web/issues/325) if anyone cares to follow along. The "Git magic" is working as expected, sufficiently hidden from normal users in GitHub and in clones.

**Comment by krishnapant13** [comment]
if it dosen't work just click on the drag and drops it will redirect you to the pc directory and simply select files from there and upload it to your repo
